Enable AutoGradMode in InferenceMode.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56107 Enable AutoGradMode in InferenceMode.**
* #56204 Move grad_mode.h/cpp to c10.

Differential Revision: [D27807137](https://our.internmc.facebook.com/intern/diff/D27807137)

Commit:
Enable AutoGradMode in InferenceMode.

[ghstack-poisoned]

Commit:
Update on "Enable AutoGradMode in InferenceMode."




[ghstack-poisoned]

Commit:
Update on "Enable AutoGradMode in InferenceMode."




[ghstack-poisoned]

bhosmer commented on c10/core/InferenceMode.h:
```diff
@@ -36,9 +40,17 @@ struct TORCH_API InferenceMode {
+  // 3. Why does setting InferenceMode also set GradMode?
+  //
+  //    This is required since InferenceMode is a faster and more restricive
+  //    version of NoGradMode. All runtime checks using GradMode::is_enabled()
```
Instead of `NoGradMode` I think it should probably say `GradMode(false)`... I'd say it's worth being fussy about the names here given how many similar ones are flying around: `GradMode`, `AutoGradMode`, `NoGradGuard`, ... there's no actual `NoGradMode`, is there? 😬 

In that same vein it might be clearer in the "[Expected TLS state in InferenceMode]" note above to say "InferenceMode(true)" and "InferenceMode(false)", or "InferenceMode enabled/disabled" instead of "InferenceMode" vs "NormalMode" - unless you're using an idiom that I'm just unfamiliar with, in which case ofc disregard 😁

albanD commented on BUILD.bazel:
```diff
@@ -83,6 +83,7 @@ cc_library(
     copts = ["-DCAFFE2_BUILD_MAIN_LIB"],
     deps = [
         ":c10_headers",
+        ":aten_headers",
```
Isn't that making a circle reference since aten depends on c10?

albanD commented on c10/core/InferenceMode.h:
```diff
@@ -11,12 +12,15 @@ struct TORCH_API InferenceMode {
+  //                  GradMode is disabled.
   //   NormalMode: InplaceOrView in raw_local_dispatch_key_set.included(),
   //               Autograd not in raw_local_dispatch_key_set.excluded()
+  //               GradMode is enabled.
```
Not necessarily :D

albanD commented on c10/core/InferenceMode.h:
```diff
@@ -61,5 +73,6 @@ struct TORCH_API InferenceMode {
   private:
     bool prev_mode;
     c10::impl::LocalDispatchKeySet prev_keyset;
+    at::AutoGradMode grad_mode;
```
nit: prev_grad_mode

ailzhang replied:
@albanD Btw `AutoGradMode` is not a state but a RAII guard (it's confusing since we have too many API in grad_mode.h) so it's actually doesn't fit the `prev_` meaning here. ;)

albanD reviewed (commented):
Could you add a test that runs a custom Function to make sure it fixes it?

malfet commented on c10/core/InferenceMode.h:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <ATen/core/grad_mode.h>
```
IMO, this `include` breaks library hierarchy, where c10 is foundatinal layer and Aten is build on top of it.
Could `grad_mode.h` be moved into c10 folder and namespace?

Commit:
Update on "Enable AutoGradMode in InferenceMode."




[ghstack-poisoned]

Commit:
Update on "Enable AutoGradMode in InferenceMode."


Differential Revision: [D27807137](https://our.internmc.facebook.com/intern/diff/D27807137)

[ghstack-poisoned]

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -534,3 +535,34 @@ TEST(InferenceModeTest, TestComplexViewInNormalMode) {
+    int mul = 2;
+    // If InferenceMode didn't set NoGradGuard automatically, this line
+    // would error out when trying to save `s` and `n` for backward.
+    auto y = MyFunction::apply(var1, mul, var2);
```
What are s and n? :D

albanD commented on c10/core/InferenceMode.h:
```diff
@@ -11,12 +12,16 @@ struct TORCH_API InferenceMode {
+  //                  GradMode is disabled.
   //   NormalMode: InplaceOrView in raw_local_dispatch_key_set.included(),
   //               Autograd not in raw_local_dispatch_key_set.excluded()
+  //               GradMode is enabled by default unless toggled manually through
```
nit: there are quite a lot of ways in which the use toggles this (that all route down to GradMode). Not sure how precise you want to be here.

albanD reviewed (approved):
LGTM
Thanks for the test!

Commit:
Update on "Enable AutoGradMode in InferenceMode."


Differential Revision: [D27807137](https://our.internmc.facebook.com/intern/diff/D27807137)

[ghstack-poisoned]
